Parse error on comment that ghc accepts

`ghcide` on current master errors out when trying to parse this:
```haskell
blademaster :: ~ » cat test.hs
main :: IO ()
main = do
  -- ^
  pure ()
```
with:
```
blademaster :: ~ » ghcide test.hs
ghcide version: 0.1.0 (GHC: 8.6.5) (PATH: /home/googleson78/.local/bin/ghcide) (GIT hash: a1cb4eb8fa27821a28b48ce05bb690d82792a18f)
Ghcide setup tester in /home/googleson78.
Report bugs at https://github.com/digital-asset/ghcide/issues

Step 1/6: Finding files to test in /home/googleson78
Found 1 files

Step 2/6: Looking for hie.yaml files that control setup
Found 1 cradle

Step 3/6, Cradle 1/1: Implicit cradle for /home/googleson78
Cradle {cradleRootDir = "/home/googleson78", cradleOptsProg = CradleAction: Default}

Step 4/6, Cradle 1/1: Loading GHC Session
Interface files cache dir: /home/googleson78/.cache/ghcide/da39a3ee5e6b4b0d3255bfef95601890afd80709

Step 5/6: Initializing the IDE

Step 6/6: Type checking the files
File:     /home/googleson78/test.hs
Hidden:   no
Range:    3:2-3:6
Source:   parser
Severity: DsError
Message:  parse error on input ‘-- ^’

Completed (1 file worked, 0 files failed)
```
But `ghc(i)` does fine:
```
blademaster :: ~ » ghc test.hs
[1 of 1] Compiling Main             ( test.hs, test.o )
Linking test ...
```
```
blademaster :: ~ » ghci test.hs
GHCi, version 8.6.5: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/googleson78/.ghci
[1 of 1] Compiling Main             ( test.hs, interpreted )
Ok, one module loaded.
>
```

My 2 cents (in case anyone cares 😄)

# Flag vs warning vs error
I agree that adding a flag for such a thing seems overkill, and not worth the cost (maintenance, bug tracking, documentation, etc.).

I personally don't mind if it is displayed as an error, but I understand some concerns/inconveniences, so displaying as a warning would be fair, too.

If I had to pick, I would keep it as an error: I prefer not to write or read code that will not be able to display documentation, it looks like a code smell to me

# Error message
That being said, I think the error message should be improved. I met this issue a couple of days ago (and opened an [issue](https://github.com/haskell/haskell-language-server/issues/148) to HLS) and it **never** occurred to me the problem was related to Haddock. 

Had I been informed it could not be parsed by Haddock, I would have solved this problem in seconds.
But because of this error message, I spent several minutes trying to understand why GHC was complaining, then several more online trying to figure out this issue.

At least when running `stack haddock` the context (lines above and below the error) help me guess it's about Haddock:
```shell
$ stack haddock
# Much noise...

Running Haddock on library for foobar-0.1.0.0..

src/Lib.hs:9:3: error: parse error on input ‘-- $’
  |
9 |   -- $
  |   ^^^^
Haddock coverage:
```

--- a/src/Development/IDE/Core/Rules.hs
+++ b/src/Development/IDE/Core/Rules.hs
@@ -29,7 +29,7 @@ module Development.IDE.Core.Rules(
 import Fingerprint
 
 import Data.Binary hiding (get, put)
-import Data.Bifunctor (first, second)
+import Data.Tuple.Extra
 import Control.Monad.Extra
 import Control.Monad.Trans.Class
 import Control.Monad.Trans.Maybe
@@ -41,7 +41,7 @@ import Development.IDE.Import.DependencyInformation
 import Development.IDE.Import.FindImports
 import           Development.IDE.Core.FileExists
 import           Development.IDE.Core.FileStore        (getFileContents)
-import           Development.IDE.Types.Diagnostics
+import           Development.IDE.Types.Diagnostics as Diag
 import Development.IDE.Types.Location
 import Development.IDE.GHC.Compat hiding (parseModule, typecheckModule)
 import Development.IDE.GHC.Util
@@ -212,7 +212,9 @@ getParsedModuleRule = defineEarlyCutoff $ \GetParsedModule file -> do
                 haddockParse = do
                     (_, (!diagsHaddock, _)) <-
                         getParsedModuleDefinition hscHaddock opt comp_pkgs file contents
-                    return diagsHaddock
+                    -- Haddock diagnostics are confusing because they don't say Haddock, and are errors.
+                    -- Fix both those properties.
+                    return $ map (\(a,b,c) -> (a,b,fixHaddockWarning c)) diagsHaddock
 
             ((fingerPrint, (diags, res)), diagsHaddock) <-
                 -- parse twice, with and without Haddocks, concurrently
@@ -220,7 +222,19 @@ getParsedModuleRule = defineEarlyCutoff $ \GetParsedModule file -> do
                 -- non-interest are always parsed with Haddocks
                 liftIO $ concurrently mainParse haddockParse
 
-            return (fingerPrint, (mergeDiagnostics diags diagsHaddock, res))
+            -- if you have a Haddock warning and a real warning at the same exact location, throw away the Haddock one
+            let realLocations = Set.fromList $ map (Diag._range . thd3) diags
+            let diagsHaddockUnique = filter (\x -> Diag._range (thd3 x) `Set.notMember` realLocations) diagsHaddock
+
+            return (fingerPrint, (diags ++ diagsHaddockUnique, res))
+
+
+-- Haddock diagnostics should be warnings and say Haddock somewhere
+fixHaddockWarning :: Diagnostic -> Diagnostic
+fixHaddockWarning x = x{_severity = Just DsWarning, _message = fixMessage $ _message x}
+  where fixMessage x | "parse error " `T.isPrefixOf` x = "Haddock " <> x
+                     | otherwise = "Haddock: " <> x
+
 
 getParsedModuleDefinition :: HscEnv -> IdeOptions -> [PackageName] -> NormalizedFilePath -> Maybe T.Text -> IO (Maybe ByteString, ([FileDiagnostic], Maybe ParsedModule))
 getParsedModuleDefinition packageState opt comp_pkgs file contents = do

--- a/src/Development/IDE/Core/Rules.hs
+++ b/src/Development/IDE/Core/Rules.hs
@@ -208,32 +208,35 @@ getParsedModuleRule = defineEarlyCutoff $ \GetParsedModule file -> do
         then
             liftIO mainParse
         else do
-            let hscHaddock = hsc{hsc_dflags = gopt_set dflags Opt_Haddock}
-                haddockParse = do
+            let haddockParse = do
                     (_, (!diagsHaddock, _)) <-
-                        getParsedModuleDefinition hscHaddock opt comp_pkgs file contents
-                    -- Haddock diagnostics are confusing because they don't say Haddock, and are errors.
-                    -- Fix both those properties.
-                    return $ map (\(a,b,c) -> (a,b,fixHaddockWarning c)) diagsHaddock
+                        getParsedModuleDefinition (withOptHaddock hsc) opt comp_pkgs file contents
+                    return diagsHaddock
 
             ((fingerPrint, (diags, res)), diagsHaddock) <-
                 -- parse twice, with and without Haddocks, concurrently
-                -- we cannot ignore Haddock parse errors because files of
-                -- non-interest are always parsed with Haddocks
+                -- we want warnings if parsing with Haddock fails
+                -- but if we parse with Haddock we lose annotations
                 liftIO $ concurrently mainParse haddockParse
 
-            -- if you have an error with Haddock and without Haddock at the same exact location, throw away the Haddock one
-            let realLocations = Set.fromList $ map (Diag._range . thd3) diags
-            let diagsHaddockUnique = filter (\x -> Diag._range (thd3 x) `Set.notMember` realLocations) diagsHaddock
-
-            return (fingerPrint, (diags ++ diagsHaddockUnique, res))
+            return (fingerPrint, (mergeParseErrorsHaddock diags diagsHaddock, res))
 
 
--- Haddock diagnostics should be warnings and say Haddock somewhere
-fixHaddockWarning :: Diagnostic -> Diagnostic
-fixHaddockWarning x = x{_severity = Just DsWarning, _message = fixMessage $ _message x}
-  where fixMessage x | "parse error " `T.isPrefixOf` x = "Haddock " <> x
-                     | otherwise = "Haddock: " <> x
+withOptHaddock :: HscEnv -> HscEnv
+withOptHaddock hsc = hsc{hsc_dflags = gopt_set (hsc_dflags hsc) Opt_Haddock}
+
+
+-- | Given some normal parse errors (first) and some from Haddock (second), merge them.
+--   Ignore Haddock errors that are in both. Demote Haddock-only errors to warnings.
+mergeParseErrorsHaddock :: [FileDiagnostic] -> [FileDiagnostic] -> [FileDiagnostic]
+mergeParseErrorsHaddock normal haddock = normal ++
+    [ (a,b,c{_severity = Just DsWarning, _message = fixMessage $ _message c})
+    | (a,b,c) <- haddock, Diag._range c `Set.notMember` locations]
+  where
+    locations = Set.fromList $ map (Diag._range . thd3) normal
+
+    fixMessage x | "parse error " `T.isPrefixOf` x = "Haddock " <> x
+                 | otherwise = "Haddock: " <> x
 
 
 getParsedModuleDefinition :: HscEnv -> IdeOptions -> [PackageName] -> NormalizedFilePath -> Maybe T.Text -> IO (Maybe ByteString, ([FileDiagnostic], Maybe ParsedModule))
@@ -645,8 +648,13 @@ getModIfaceRule = define $ \GetModIface f -> do
             opt <- getIdeOptions
             (_, contents) <- getFileContents f
             -- Embed --haddocks in the interface file
-            hsc <- pure hsc{hsc_dflags = gopt_set (hsc_dflags hsc) Opt_Haddock}
-            (_, (diags, mb_pm)) <- liftIO $ getParsedModuleDefinition hsc opt comp_pkgs f contents
+            (_, (diags, mb_pm)) <- liftIO $ getParsedModuleDefinition (withOptHaddock hsc) opt comp_pkgs f contents
+            (diags, mb_pm) <- case mb_pm of
+                Just _ -> return (diags, mb_pm)
+                Nothing -> do
+                    -- if parsing fails, try parsing again with Haddock turned off
+                    (_, (diagsNoHaddock, mb_pm)) <- liftIO $ getParsedModuleDefinition hsc opt comp_pkgs f contents
+                    return (mergeParseErrorsHaddock diagsNoHaddock diags, mb_pm)
             case mb_pm of
                 Nothing -> return (diags, Nothing)
                 Just pm -> do

--- a/src/Development/IDE/Core/Rules.hs
+++ b/src/Development/IDE/Core/Rules.hs
@@ -222,7 +222,7 @@ getParsedModuleRule = defineEarlyCutoff $ \GetParsedModule file -> do
                 -- non-interest are always parsed with Haddocks
                 liftIO $ concurrently mainParse haddockParse
 
-            -- if you have a Haddock warning and a real warning at the same exact location, throw away the Haddock one
+            -- if you have an error with Haddock and without Haddock at the same exact location, throw away the Haddock one
             let realLocations = Set.fromList $ map (Diag._range . thd3) diags
             let diagsHaddockUnique = filter (\x -> Diag._range (thd3 x) `Set.notMember` realLocations) diagsHaddock
 

--- a/src/Development/IDE/GHC/Error.hs
+++ b/src/Development/IDE/GHC/Error.hs
@@ -9,7 +9,6 @@ module Development.IDE.GHC.Error
   , diagFromStrings
   , diagFromGhcException
   , catchSrcErrors
-  , mergeDiagnostics
 
   -- * utilities working with spans
   , srcSpanToLocation
@@ -63,26 +62,6 @@ diagFromErrMsg diagSource dflags e =
 
 diagFromErrMsgs :: T.Text -> DynFlags -> Bag ErrMsg -> [FileDiagnostic]
 diagFromErrMsgs diagSource dflags = concatMap (diagFromErrMsg diagSource dflags) . bagToList
-
--- | Merges two sorted lists of diagnostics, removing duplicates.
---   Assumes all the diagnostics are for the same file.
-mergeDiagnostics :: [FileDiagnostic] -> [FileDiagnostic] -> [FileDiagnostic]
-mergeDiagnostics aa [] = aa
-mergeDiagnostics [] bb = bb
-mergeDiagnostics (a@(_,_,ad@Diagnostic{_range = ar}):aa) (b@(_,_,bd@Diagnostic{_range=br}):bb)
-  | ar < br
-  = a : mergeDiagnostics aa (b:bb)
-  | br < ar
-  = b : mergeDiagnostics (a:aa) bb
-  | _severity ad == _severity bd
-  && _source ad == _source bd
-  && _message ad == _message bd
-  && _code ad == _code bd
-  && _relatedInformation ad == _relatedInformation bd
-  && _tags ad == _tags bd
-  = a : mergeDiagnostics aa bb
-  | otherwise
-  = a : b : mergeDiagnostics aa bb
 
 -- | Convert a GHC SrcSpan to a DAML compiler Range
 srcSpanToRange :: SrcSpan -> Range

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -433,7 +433,7 @@ diagnosticTests = testGroup "diagnostics"
       _ <- createDoc "Foo.hs" "haskell" fooContent
       expectDiagnostics
         [ ( "Foo.hs"
-          , [(DsError, (2, 8), "Parse error on input")
+          , [(DsWarning, (2, 8), "Haddock parse error on input")
             ]
           )
         ]

